truffle/src/main/java/org/jruby/truffle/language/constants/LookupConstantNode.java

```diff
@@ -74,22 +77,35 @@ protected RubyConstant lookupConstant(
         if (!isVisible) {
             throw new RaiseException(coreExceptions().nameErrorPrivateConstant(module, name, this));
         }
+        if (constant != null && constant.isDeprecated()) {
+            warnDeprecatedConstant(frame, name);
+        }
         return constant;
     }
 
+    private void warnDeprecatedConstant(VirtualFrame frame, String name) {
+        if (warnNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            warnNode = insert(new WarnNode());
+        }
+        warnNode.execute(frame, "constant " , name , " is deprecated");
+    }
+
     public Assumption getUnmodifiedAssumption(DynamicObject module) {
         return Layouts.MODULE.getFields(module).getUnmodifiedAssumption();
     }
 
-    @TruffleBoundary
```

Comment:
@eregon I've removed a few `TruffleBoundary`s  here to be able to pass the `VirtualFrame` parameter. Does something in these methods need to be behind a boundary? Thanks.